File: 백준/Gold/2493. 탑/탑.java
import java.util.*;
import java.io.*;

public class Main {
    public static void main(String[] args) throws IOException {
        Stack<Integer> stack = new Stack<>();

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        StringBuilder sb = new StringBuilder();

        int n = Integer.parseInt(st.nextToken());
        int[] towers = new int[n];

        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < n; i++) {
            towers[i] = Integer.parseInt(st.nextToken());
        }

        int[] result = new int[n];

        for (int i = 0; i < n; i++) {
            int num = towers[i];

            while (!stack.isEmpty()) {
                if (towers[stack.peek()] > num) {
                    result[i] = stack.peek() + 1;
                    break;
                } else {
                    stack.pop();
                }
            }
            stack.push(i);
        }
        for (int e: result) {
            sb.append(e);
            sb.append(" ");
        }

        System.out.println(sb);
    }
}

File: 백준/Gold/2493. 탑/탑.py
# 탑
import sys
input = sys.stdin.readline

n = int(input())
towers = list(map(int,input().split()))
MAX = 100000001
result = [0] * (n+1)
result[0] = MAX

stack = [(0, MAX), (1, towers[0])]

for i in range(1, n):
    if towers[i] > stack[-1][1]:
        while stack and stack[-1][1] < towers[i]:
            idx, num = stack.pop()
            result[idx] = stack[-1][0]

    stack.append((i+1, towers[i]))
    
            
while len(stack) > 1:
    idx, num = stack.pop()
    result[idx] = stack[-1][0]

print(*result[1:])
